File: cmd/loom-task-image-builder-supervisor/publication_status.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"unicode/utf8"
)

const maxPublicationStatusBytes = 4096

type publicationStatusBinding struct {
	GrantID              string `json:"grant_id"`
	OperationID          string `json:"operation_id"`
	MaterializationID    string `json:"materialization_id"`
	AttemptID            string `json:"attempt_id"`
	LeaseEpoch           int64  `json:"lease_epoch"`
	PinnedSnapshotSHA256 string `json:"pinned_snapshot_sha256"`
	CandidateSetSHA256   string `json:"candidate_set_sha256"`
	ComponentCount       int    `json:"component_count"`
}

type publicationStatus struct {
	GrantID            string
	OperationID        string
	MaterializationID  string
	AttemptID          string
	LeaseEpoch         int64
	State              string
	SnapshotSHA256     string
	CandidateSetSHA256 string
	ComponentCount     int
	Receipt            *publicationReceipt
	FailureCode        string
}

// Field order is RFC 8785 lexical order for this closed ASCII-only schema.
type publicationStatusWire struct {
	AttemptID          *string         `json:"attempt_id"`
	CandidateSetSHA256 *string         `json:"candidate_set_sha256"`
	ComponentCount     *int            `json:"component_count"`
	FailureCode        *string         `json:"failure_code,omitempty"`
	GrantID            *string         `json:"grant_id"`
	LeaseEpoch         *int64          `json:"lease_epoch"`
	MaterializationID  *string         `json:"materialization_id"`
	OperationID        *string         `json:"operation_id"`
	Receipt            json.RawMessage `json:"receipt,omitempty"`
	Schema             *string         `json:"schema"`
	SnapshotSHA256     *string         `json:"snapshot_sha256"`
	State              *string         `json:"state"`
}

func parsePublicationStatus(payload []byte, binding publicationStatusBinding) (publicationStatus, error) {
	invalid := func() (publicationStatus, error) {
		return publicationStatus{}, errors.New("publication status invalid")
	}
	if len(payload) == 0 || len(payload) > maxPublicationStatusBytes || !utf8.Valid(payload) || !validPublicationStatusBinding(binding) {
		return invalid()
	}
	var wire publicationStatusWire
	if err := decodeStrictJSON(payload, &wire); err != nil ||
		wire.AttemptID == nil ||
		wire.CandidateSetSHA256 == nil ||
		wire.ComponentCount == nil ||
		wire.GrantID == nil ||
		wire.LeaseEpoch == nil ||
		wire.MaterializationID == nil ||
		wire.OperationID == nil ||
		wire.Schema == nil ||
		wire.SnapshotSHA256 == nil ||
		wire.State == nil {
		return invalid()
	}
	if *wire.Schema != "loom.task-image-publication-status/v1" ||
		!isCanonicalNonZeroUUID(*wire.GrantID) ||
		!isCanonicalNonZeroUUID(*wire.OperationID) ||
		!isCanonicalNonZeroUUID(*wire.MaterializationID) ||
		!isCanonicalNonZeroUUID(*wire.AttemptID) ||
		!isPositiveJSONSafeInteger(*wire.LeaseEpoch) ||
		!isDigest(*wire.SnapshotSHA256) ||
		!isDigest(*wire.CandidateSetSHA256) ||
		*wire.ComponentCount < 1 ||
		*wire.ComponentCount > maxPublicationComponents ||
		!validPublicationStatusTerminalShape(*wire.State, wire.Receipt, wire.FailureCode) {
		return invalid()
	}
	canonical, err := json.Marshal(wire)
	if err != nil || !bytes.Equal(canonical, payload) {
		return invalid()
	}
	if *wire.GrantID != binding.GrantID ||
		*wire.OperationID != binding.OperationID ||
		*wire.MaterializationID != binding.MaterializationID ||
		*wire.AttemptID != binding.AttemptID ||
		*wire.LeaseEpoch != binding.LeaseEpoch ||
		(binding.PinnedSnapshotSHA256 != "" && *wire.SnapshotSHA256 != binding.PinnedSnapshotSHA256) ||
		*wire.CandidateSetSHA256 != binding.CandidateSetSHA256 ||
		*wire.ComponentCount != binding.ComponentCount {
		return invalid()
	}

	status := publicationStatus{
		GrantID:            *wire.GrantID,
		OperationID:        *wire.OperationID,
		MaterializationID:  *wire.MaterializationID,
		AttemptID:          *wire.AttemptID,
		LeaseEpoch:         *wire.LeaseEpoch,
		State:              *wire.State,
		SnapshotSHA256:     *wire.SnapshotSHA256,
		CandidateSetSHA256: *wire.CandidateSetSHA256,
		ComponentCount:     *wire.ComponentCount,
	}
	if wire.FailureCode != nil {
		status.FailureCode = *wire.FailureCode
	}
	if len(wire.Receipt) != 0 {
		receipt, err := parsePublicationReceipt(wire.Receipt, publicationReceiptBinding{
			OperationID:        status.OperationID,
			MaterializationID:  status.MaterializationID,
			AttemptID:          status.AttemptID,
			LeaseEpoch:         status.LeaseEpoch,
			SnapshotSHA256:     status.SnapshotSHA256,
			CandidateSetSHA256: status.CandidateSetSHA256,
			ComponentCount:     status.ComponentCount,
		})
		if err != nil {
			return invalid()
		}
		status.Receipt = &receipt
	}
	return status, nil
}

func validPublicationStatusBinding(binding publicationStatusBinding) bool {
	return isCanonicalNonZeroUUID(binding.GrantID) &&
		isCanonicalNonZeroUUID(binding.OperationID) &&
		isCanonicalNonZeroUUID(binding.MaterializationID) &&
		isCanonicalNonZeroUUID(binding.AttemptID) &&
		isPositiveJSONSafeInteger(binding.LeaseEpoch) &&
		(binding.PinnedSnapshotSHA256 == "" || isDigest(binding.PinnedSnapshotSHA256)) &&
		isDigest(binding.CandidateSetSHA256) &&
		binding.ComponentCount >= 1 &&
		binding.ComponentCount <= maxPublicationComponents
}

func validPublicationStatusTerminalShape(state string, receipt json.RawMessage, failureCode *string) bool {
	hasReceipt := len(receipt) != 0
	hasFailure := failureCode != nil
	switch state {
	case "queued", "running":
		return !hasReceipt && !hasFailure
	case "failed":
		return !hasReceipt && hasFailure && isPublicationFailureCode(*failureCode)
	case "completed":
		return hasReceipt && !hasFailure
	default:
		return false
	}
}

func isPublicationFailureCode(value string) bool {
	switch value {
	case "integrity", "authority_lost", "verification_failed", "deadline":
		return true
	default:
		return false
	}
}

File: src/loom_task_image_authority/publication_status.py
"""Small durable-job projections for later authenticated fixed submit/poll APIs.

Projection is not authentication, signature verification or database completion
proof. Callers must authenticate the current session, authorize the exact job,
and verify any historical receipt through the completion store before projecting.
Never serialize PublicationJob directly into an HTTP or guard polling response.
"""

from __future__ import annotations

from typing import Annotated, Literal

import rfc8785
from pydantic import Field, model_validator

from loom_task_image_authority.contracts import Digest
from loom_task_image_authority.publication_contracts import (
    CanonicalUUID,
    SafePositiveInteger,
    _ClosedPublicationModel,
    _decode,
)
from loom_task_image_authority.publication_jobs import (
    PublicationFailureCode,
    PublicationJob,
    canonical_snapshot_bytes,
    decode_publication_snapshot,
)
from loom_task_image_authority.publication_receipts import (
    PublicationCandidateIdentity,
    PublicationReceipt,
    candidate_set_sha256,
    canonical_receipt_bytes,
    decode_publication_receipt,
)

MAX_STATUS_BYTES = 4096


class PublicationStatus(_ClosedPublicationModel):
    schema_name: Literal["loom.task-image-publication-status/v1"] = Field(alias="schema")
    grant_id: CanonicalUUID
    operation_id: CanonicalUUID
    materialization_id: CanonicalUUID
    attempt_id: CanonicalUUID
    lease_epoch: SafePositiveInteger
    state: Literal["queued", "running", "failed", "completed"]
    snapshot_sha256: Digest
    candidate_set_sha256: Digest
    component_count: Annotated[int, Field(strict=True, ge=1, le=128)]
    receipt: PublicationReceipt | None = None
    failure_code: PublicationFailureCode | None = None

    @model_validator(mode="after")
    def _terminal_binding(self) -> PublicationStatus:
        if (self.state == "completed") != (self.receipt is not None) or (
            self.state == "failed"
        ) != (self.failure_code is not None):
            raise ValueError("publication status terminal shape mismatch")
        if self.receipt is not None:
            for field in (
                "operation_id",
                "materialization_id",
                "attempt_id",
                "lease_epoch",
                "snapshot_sha256",
                "candidate_set_sha256",
                "component_count",
            ):
                if getattr(self.receipt, field) != getattr(self, field):
                    raise ValueError("publication status receipt binding mismatch")
        return self


def decode_publication_status(encoded: bytes) -> PublicationStatus:
    return _decode(encoded, PublicationStatus, MAX_STATUS_BYTES)


def canonical_status_bytes(status: PublicationStatus) -> bytes:
    checked = PublicationStatus.model_validate_json(
        status.model_dump_json(by_alias=True, exclude_none=True)
    )
    encoded = rfc8785.dumps(checked.model_dump(mode="json", by_alias=True, exclude_none=True))
    if len(encoded) > MAX_STATUS_BYTES:
        raise ValueError("publication status exceeds byte ceiling")
    return encoded


def project_publication_status(
    job: PublicationJob, *, receipt: PublicationReceipt | None = None
) -> PublicationStatus:
    """Own validated nonsecret scalar bindings; never return frozen graph inputs."""
    # Job times are internal aware datetimes, not this response's wire fields.
    # Reparse the separate canonical snapshot while preserving those strict types.
    fields = job.model_dump(by_alias=True, exclude_none=True, exclude={"snapshot"})
    fields["snapshot"] = decode_publication_snapshot(canonical_snapshot_bytes(job.snapshot))
    checked = PublicationJob.model_validate(fields)
    snapshot = checked.snapshot
    identity_hash = candidate_set_sha256(
        tuple(
            PublicationCandidateIdentity(
                candidate_id=str(item.candidate.candidate_id), component=item.candidate.component
            )
            for item in snapshot.components
        )
    )
    values: dict[str, object] = dict(
        schema="loom.task-image-publication-status/v1",
        grant_id=snapshot.grant_id,
        operation_id=checked.operation_id,
        materialization_id=snapshot.materialization_id,
        attempt_id=snapshot.attempt_id,
        lease_epoch=snapshot.lease_epoch,
        state=checked.state,
        snapshot_sha256=checked.snapshot_sha256,
        candidate_set_sha256=identity_hash,
        component_count=len(snapshot.components),
    )
    if receipt is not None:
        owned_receipt = decode_publication_receipt(canonical_receipt_bytes(receipt))
        if owned_receipt.worker_generation != checked.worker_generation:
            raise ValueError("publication status worker generation mismatch")
        values["receipt"] = owned_receipt
    if checked.failure_code is not None:
        values["failure_code"] = checked.failure_code
    return PublicationStatus.model_validate(values)
